File: library/src/main/java/wannabe/grid/iterators/TranslatingIterator.java
package wannabe.grid.iterators;

import java.util.Iterator;
import wannabe.Translation;
import wannabe.Voxel;

public final class TranslatingIterator implements Iterator<Voxel> {
  private final Iterator<Voxel> realIterator;
  private final Translation translation;
  /** Just a reused object to do math for us. */
  private final Translation workhorse = new Translation(0, 0, 0);

  public TranslatingIterator(Iterator<Voxel> realIterator, Translation translation) {
    this.realIterator = realIterator;
    this.translation = translation;
    if (translation.isZero()) {
      throw new IllegalStateException("Don't use a translatingIterator unless you need to!");
    }
  }

  @Override public boolean hasNext() {
    return realIterator.hasNext();
  }

  @Override public Voxel next() {
    Voxel real = realIterator.next();
    return new Voxel(workhorse.set(real.position).add(translation).asPosition(), real.value);
  }

  @Override public void remove() {
    // TODO should I support remove?
    realIterator.remove();
  }
}

File: library/src/main/java/wannabe/grid/iterators/TranslatingIterator.kt
package wannabe.grid.iterators

import wannabe.Translation
import wannabe.Voxel

class TranslatingIterator(
  private val realIterator: Iterator<Voxel>,
  private val translation: Translation
) : Iterator<Voxel> {

  /** Just a reused object to do math for us.  */
  private val workhorse = Translation(0, 0, 0)

  init {
    check(!translation.isZero) { "Don't use a translatingIterator unless you need to!" }
  }

  override fun hasNext(): Boolean = realIterator.hasNext()

  override fun next(): Voxel {
    val (position, value) = realIterator.next()
    return Voxel(
        workhorse.set(position)
            .add(translation)
            .asPosition(), value
    )
  }
}
